R/prepare_stop_level_data.R

```diff
@@ -26,146 +29,347 @@ granulate_to_stop <- function(mbbs) {
+        #additional checks go here.
+      )
+      out # return
+    })
+  
+}
+
+#' Use when species comments follows 1=5 or st 1=5 to stop 1=5 format
+#' Takes a string and returns a list of 20 numbers
+#' @param x a string from the species_comment column formatted by "stop=count"
+#' @param count an integer from the count column of the mbbs
+#' @importFrom stringr str_split_1 str_detect str_replace_all
+#' @importFrom assertthat assert_that
+#' @returns out a standardized list of 20 integers, representing the 
+#'    species count at 20 stops.
+sp_com_stop_equals_count <- function(x, count = -99) {
+  
+  assertthat::assert_that(
+    str_detect(x, "=") # check has equals as format requires
+  )
+  
+  # split based , or ;
+  x %>% 
+    stringr::str_split_1(",|;") %>%
+    str_replace_all(
+      c( # Convert empty strings 0
+        "^$" = "0",
+        # Remove the Stops = / Stops:
+        "Stop(s)?( )?(=)?(:)?( )?" = "",
+        # If it starts with st, drop
+        "( )?st(op)?( )?" = ""
+      )) %>%
+    # make standard list of 20
+    standardize_stops_equals() %>%
+    (\(out) {
+      assertthat::assert_that(
+        length(out) == 20
+        #sum(out) == count #this check is no longer done here.
+        #additional checks go here.
+      )
+      out # return
+    })
+}
+
+
+#' Use when species comment is ie: "Stop 6"
+#' The count for that stop is the count for the whole route
+#' and the species was not seen at any other stops
+#' @param x a string from the species_comment column formatted by "stop=count"
+#' @param count an integer from the count column of the mbbs
+#' @importFrom stringr str_count
+sp_com_only_stop <- function(x, count = -99) {
+  
+  assertthat::assert_that(
+    # check there's only one number, up to two digits
+    str_count(x, "[0-9]([0-9])?") == 1,
+    # count has been given
+    count > 0 
+  )
+  
+  x %>%
+    # remove anything after a comma (ie: descriptive notes)
+    str_extract(".*(,)?") %>%
+    # extract numbers
+    str_extract("[0-9]([0-9])?") %>%
+    as.numeric() %>%
+    # make standard list of 20
+    paste("=", count, sep = "") %>%
+    standardize_stops_equals() %>%
+    (\(out) {
+      assertthat::assert_that(
+        length(out) == 20
+        #sum(out) == count #this check is no longer done here.
+        # additional checks go here.
+      )
+      out # return
+    })
+}
+
+#' Prepare mbbs dataset for processing species comments
+#' @importFrom dplyr filter mutate bind_cols tibble relocate
+#' @importFrom rlang set_names
+#' @importFrom stringr str_detect
+prepare_to_process <- \(mbbs) {
+  mbbs %>%
+    # Keep rows that have non-blank species_comments
+    filter(species_comments != "") %>%
+    # Keep rows that aren't already separated by stop
+    filter(is.na(stop_num)) %>%
+    # Keep rows that contain at least one number
+    filter(str_detect(species_comments, "[0-9]")) %>%
+    # Remove pre-dawn owling checklist
+    filter(sub_id != "S46296942") %>%
```

Comment:
Should this removal go somewhere else? Feels weird to have this hard-coded here.

Reply:
It should go elsewhere ultimately, but can I hardcode it here and leave a note in #17 about removing it? Part of the issue is this checklist has only three stops but gets caught in the filter of checking for species_comments. I don't think I can pull it out without referencing the sub_id - otherwise I suppose I'd filter for species comments that are, in their entirety, ",1,1" ",1," and "1,1,1" but this is more targeted. Part of this is the check at the end of the sp_com functions that checks the length of the output is 20

Reply:
Yeah, let's be sure to open an issue to resolve this later. I'd like to be sure all hardcoded removals are in one spot.

Reply:
Added to the identify pre-dawn owling checklists issue

Reply:
Oh, sorry I see - it should be part of the import_data process if there's any hardcoded removals there. I will open a seperate issue for it.